Use geth internal GetProof for monomer eth_getProof

natebeauregard opened the pull request:
### Background
Monomer was previously using a lightly modifying version of geth's internal `GetProof` func for the monomer `eth_getProof` endpoint. To minimize duplicated code and the potential for errors, the original geth `GetProof` func was copied to monomer's internal `ethapi` package and all monomer specific logic is now included in a custom monomer ETH API Backend.

### Changes
- Use geth internal GetProof for `eth_getProof`
  - Copies geth `eth_getProof` impl and helper funcs into the `ethapi` package
  - Adds a monomer eth API backend to handle monomer specific logic when calling the internal `eth_getProof` RPC method
  - Removes `get_proof.go` and call into the internal geth API for `eth_getProof` from eth.go
- Rename eth API helpers to include the API suffix

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
	- Introduced `EthAPIBackend` for enhanced Ethereum API interactions, including block header and state retrieval.
	- Added `BlockChainAPI` for blockchain data retrieval and proof generation.
	- New `ProofAPI` type to manage blockchain proofs.

- **Improvements**
	- Renamed types and functions for clarity, transitioning to an API-centric design.
	- Enhanced error handling and modularity in the codebase.

- **Testing Enhancements**
	- Updated test cases to utilize new API constructors, improving consistency and maintainability.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

joshklop commented on eth/api_backend.go:
```diff
@@ -0,0 +1,93 @@
+	return header.ToEth(), nil
+}
+
+func (*EthAPIBackend) HistoricalRPCService() *rpc.Client {
+	return nil
+}
```
Why do we need this?

natebeauregard replied:
We'll never hit it because our chain config will always be post bedrock, however it's used [here](https://github.com/ethereum-optimism/op-geth/blob/3653ceb/internal/ethapi/api.go#L713-L724) in the op-geth impl.

joshklop commented on eth/api_backend.go:
```diff
@@ -0,0 +1,93 @@
+	"github.com/polymerdao/monomer/eth/internal/ethapi"
+)
+
+type EthAPIBackend struct {
```
can this be un-exported?

joshklop commented on eth/api_backend.go:
```diff
@@ -0,0 +1,93 @@
+	blockStore DB
+}
+
+func NewEthAPIBackend(db state.Database, blockStore DB) *EthAPIBackend {
```
same, un-exported

joshklop reviewed (approved):
Approving pending a few minor fixes, and assuming `internal/ethapi/api.go` continues to match the geth implementation 1:1

natebeauregard commented:
@joshklop I addressed your code review comments and included them in the first commit before I pushed back up. I ran a diff of our copied op-geth api and the original op-geth api at commit 3653ceb and reordered some of the pre-existing api funcs we had to match the ordering from the op-geth api. I also added an additional commit to update the README link to redirect to the correct op-geth commit.

joshklop commented on eth/eth.go:
```diff
@@ -88,3 +92,23 @@ func (e *Block) toRPCBlock(block *monomer.Block, fullTx bool) (map[string]any, e
+// GetProof returns the account and storage values of the specified account including the Merkle-proof.
+// The user can specify either a block number or a block hash to build to proof from.
+func (p *ProofProvider) GetProof(
+	address common.Address,
```
We should accept a `context.Context` here as well (it is permissible with the geth RPC server).

natebeauregard replied:
Good catch, I'll add that in the first commit and then merge
